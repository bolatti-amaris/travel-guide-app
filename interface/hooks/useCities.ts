import { useEffect, useCallback } from 'react';
import { City } from '@/core/domain/entities/City';
import { useCitiesStore } from '../../core/interface/store/useCitiesStore';

export const useCities = () => {
  const {
    cities,
    isLoading,
    error,
    fetchCities,
    selectedCity,
    setSelectedCity,
  } = useCitiesStore();

  useEffect(() => {
    fetchCities();
  }, []);

  const selectCity = (city: City) => {
    setSelectedCity(city);
  };

  const reset = useCallback(() => {
    useCitiesStore.getState().reset();
  }, []);

  return {
    cities,
    isLoading,
    error,
    selectedCity,
    selectCity,
    reset,
  };
};
