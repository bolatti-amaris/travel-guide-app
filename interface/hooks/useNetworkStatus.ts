import NetInfo from '@react-native-community/netinfo';
import { useEffect, useState } from 'react';

export const useNetworkStatus = () => {
  const [isConnected, setIsConnected] = useState(true);

  useEffect(() => {
    const unsubscribe = NetInfo.addEventListener((state) => {
      setIsConnected(Boolean(state.isConnected));
    });

    return () => unsubscribe();
  }, []);

  return { isConnected };
};
