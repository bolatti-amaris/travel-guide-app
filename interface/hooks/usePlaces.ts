import { useCallback, useEffect } from 'react';
import { useCitiesStore } from '@/core/interface/store/useCitiesStore';
import { usePlacesStore } from '@/core/interface/store/usePlacesStore';

export const usePlaces = () => {
  const { selectedCity } = useCitiesStore();
  const { places, isLoading, error, fetchPlaces } = usePlacesStore();

  useEffect(() => {
    const shouldFetch =
      selectedCity &&
      selectedCity.key &&
      places.length === 0 &&
      !isLoading &&
      !error;

    if (shouldFetch) {
      fetchPlaces(selectedCity.key);
    }
  }, [selectedCity, places.length, isLoading, error, fetchPlaces]);

  const reset = useCallback(() => {
    usePlacesStore.getState().reset();
  }, []);

  return {
    places,
    isLoading,
    error,
    fetchPlaces,
    reset,
  };
};
