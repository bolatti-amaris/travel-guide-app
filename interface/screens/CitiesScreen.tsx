import { useCities } from '@/interface/hooks/useCities';
import CityList from '@/interface/components/CityList';
import React from 'react';
import { View, Text } from 'react-native';
import Loader from '../shared/Loader';
import Retry from '../shared/Retry';

const CitiesScreen = () => {
    const { cities, isLoading, error, fetchCities } = useCities();

    const showEmptyState = !isLoading && !error && cities.length === 0;

  return (
    <View className="flex-1 bg-gray-100 px-4 py-6">
        {isLoading && (
        <Loader message="Loading cities..." />
        )}

        {error && (
        <Retry
        message={error}
        onRetry={fetchCities}
        />
        )}

        {showEmptyState && (
        <Retry
            message="No cities available"
            onRetry={fetchCities}
        />
        )}

        {!isLoading && !error && cities.length > 0 && (
        <CityList cities={cities} />
        )}
    </View>
  );
};

export default CitiesScreen;