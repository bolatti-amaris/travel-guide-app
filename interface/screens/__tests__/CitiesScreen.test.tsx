import React from 'react';
import { render } from '@testing-library/react-native';
import CitiesScreen from '../CitiesScreen';

import { useCities } from '../../hooks/useCities';

jest.mock('expo-router', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
    setOptions: jest.fn(),
  }),
}));

jest.mock('../../hooks/useCities', () => ({
  useCities: jest.fn(),
}));

jest.mock('../../shared/Loader', () => {
  const { View, Text } = require('react-native');
  return () => (
    <View testID="loader">
      <Text>Loader</Text>
    </View>
  );
});

jest.mock('../../shared/Retry', () => {
  const { View, Text } = require('react-native');
  return ({ message }: { message: string }) => (
    <View testID={message === 'No cities available' ? 'empty-retry' : 'retry'}>
      <Text>{message}</Text>
    </View>
  );
});

jest.mock('../../components/CityList', () => {
  const { View, Text } = require('react-native');
  return ({ cities }: { cities: any[] }) => (
    <View testID="city-list">
      <Text>{cities.length} cities</Text>
    </View>
  );
});

describe('CitiesScreen', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('shows loader', () => {
    (useCities as jest.Mock).mockReturnValue({
      cities: [],
      isLoading: true,
      error: null,
      fetchCities: jest.fn(),
      reset: jest.fn(),
    });

    const { getByTestId } = render(<CitiesScreen />);
    expect(getByTestId('loader')).toBeTruthy();
  });

  it('shows retry on error', () => {
    (useCities as jest.Mock).mockReturnValue({
      cities: [],
      isLoading: false,
      error: 'Failed to fetch',
      fetchCities: jest.fn(),
      reset: jest.fn(),
    });

    const { getByTestId } = render(<CitiesScreen />);
    expect(getByTestId('retry')).toBeTruthy();
  });

  it('shows retry when empty state', () => {
    (useCities as jest.Mock).mockReturnValue({
      cities: [],
      isLoading: false,
      error: null,
      fetchCities: jest.fn(),
      reset: jest.fn(),
    });

    const { getByTestId } = render(<CitiesScreen />);
    expect(getByTestId('empty-retry')).toBeTruthy();
  });

  it('shows CityList when cities are loaded', () => {
    (useCities as jest.Mock).mockReturnValue({
      cities: [
        {
          id: 1,
          key: 'paris',
          name: 'Paris',
          nativeName: 'Paris',
          currency: 'EUR',
          language: 'fr',
        },
      ],
      isLoading: false,
      error: null,
      fetchCities: jest.fn(),
      reset: jest.fn(),
    });

    const { getByTestId } = render(<CitiesScreen />);
    expect(getByTestId('city-list')).toBeTruthy();
  });

  it('matches snapshot', () => {
    (useCities as jest.Mock).mockReturnValue({
      cities: [],
      isLoading: false,
      error: null,
      fetchCities: jest.fn(),
      reset: jest.fn(),
    });

    const tree = render(<CitiesScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
