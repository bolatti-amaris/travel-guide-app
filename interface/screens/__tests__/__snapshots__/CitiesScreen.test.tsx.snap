// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CitiesScreen matches snapshot 1`] = `
<View
  className="flex-1 px-4 bg-background dark:bg-background-dark"
  testID="cities-screen"
>
  <View
    testID="empty-retry"
  >
    <Text>
      No cities available
    </Text>
  </View>
</View>
`;
